### ISSUE TITLE
bug: performance regression in `2ca076e45` - 100% CPU and lag on 100 LOC file

### ISSUE THREAD

**Issue by llllvvuu** [neovim/neovim#25074]
EDIT: I found the offending edit. PR opened: #25075

### Problem

When typing on https://github.com/neovim/neovim/commit/2ca076e45fb3f1c08f6a1a374834df0701b8d778 or later, there is significant latency and CPU spike to 100%. When typing in the same file on any commit before https://github.com/neovim/neovim/commit/2ca076e45fb3f1c08f6a1a374834df0701b8d778, there is no latency and the CPU spikes to 3% at most.

Note:
* `cursorcolumn` isn't necessary to reproduce the issue, but if you turn it on it's much more clear (otherwise it's only noticeable on some keystrokes).
* Needs both highlight and indent enabled, won't reproduce without one of them disabled.

Pinging author @lewis6991 for any insight.

### UPDATE: Found blowup in `iter_captures`:
```diff
diff --git a/runtime/lua/vim/treesitter/query.lua b/runtime/lua/vim/treesitter/query.lua
index 309365731..0f8b2502f 100644
--- a/runtime/lua/vim/treesitter/query.lua
+++ b/runtime/lua/vim/treesitter/query.lua
@@ -680,6 +680,7 @@ local function value_or_node_range(start, stop, node)
   return start, stop
 end
 
+local my_count = 0
 --- Iterate over all captures from all matches inside {node}
 ---
 --- {source} is needed if the query contains predicates; then the caller
@@ -716,7 +717,10 @@ function Query:iter_captures(node, source, start, stop)
   start, stop = value_or_node_range(start, stop, node)
 
   local raw_iter = node:_rawquery(self.query, true, start, stop)
-  local function iter()
+  local function iter(depth)
+    my_count = my_count + 1
+    print("my_count", my_count)
+    print("my_depth", depth)
     local capture, captured_node, match = raw_iter()
     local metadata = {}
 
@@ -724,7 +728,7 @@ function Query:iter_captures(node, source, start, stop)
       local active = self:match_preds(match, match.pattern, source)
       match.active = active
       if not active then
-        return iter() -- tail call: try next match
+        return iter((depth or 0) + 1) -- tail call: try next match
       end
 
       self:apply_directives(match, match.pattern, source, metadata)
```

After typing a single char (`nvim -u init.lua` w/ provided file, then `Of<Esc>:messages`):

| after #24647 | before #24647 |
| --- | --- |
| my_depth 2618 | my_depth 5 |
| my_count 24966 | my_count 1639 |

Adding:
```lua
    if depth and depth > 2000 then
      if captured_node then
        a, b, c, d = captured_node:range()
        print(a, b, c, d)
      end
      -- print(self.dbgquery)
    end
```
Shows that every single token is being matched (and many multiple times):
```
96 20 96 23
96 24 96 36
96 24 96 36
96 24 96 36
96 24 96 36
96 24 96 36
96 36 96 37
96 36 96 37
96 36 96 37
96 36 96 37
96 36 96 37
```

for the query:
```lisp
(_) @spell
((tag
  (name) @text.todo @nospell
  ("(" @punctuation.bracket (user) @constant ")" @punctuation.bracket)?
  ":" @punctuation.delimiter)
  (#any-of? @text.todo "TODO" "WIP"))
("text" @text.todo @nospell
 (#any-of? @text.todo "TODO" "WIP"))
((tag
  (name) @text.note @nospell
  ("(" @punctuation.bracket (user) @constant ")" @punctuation.bracket)?
  ":" @punctuation.delimiter)
  (#any-of? @text.note "NOTE" "XXX" "INFO" "DOCS" "PERF" "TEST"))
("text" @text.note @nospell
 (#any-of? @text.note "NOTE" "XXX" "INFO" "DOCS" "PERF" "TEST"))
((tag
  (name) @text.warning @nospell
  ("(" @punctuation.bracket (user) @constant ")" @punctuation.bracket)?
  ":" @punctuation.delimiter)
  (#any-of? @text.warning "HACK" "WARNING" "WARN" "FIX"))
("text" @text.warning @nospell
 (#any-of? @text.warning "HACK" "WARNING" "WARN" "FIX"))
((tag
  (name) @text.danger @nospell
  ("(" @punctuation.bracket (user) @constant ")" @punctuation.bracket)?
  ":" @punctuation.delimiter)
  (#any-of? @text.danger "FIXME" "BUG" "ERROR"))
("text" @text.danger @nospell
 (#any-of? @text.danger "FIXME" "BUG" "ERROR"))
; Issue number (#123)
("text" @number
 (#lua-match? @number "^#[0-9]+$"))
((uri) @text.uri @nospell)

```

^ this query shouldn't even be running on the source code; it's for comments only.

### Flame graphs (download and open locally for interactive):

Before #24647
![flame](https://github.com/neovim/neovim/assets/5601392/65ee5f4b-54d4-4ce1-a4a3-da1e559a9522)
After #24647
![flame2](https://github.com/neovim/neovim/assets/5601392/44af8a26-20ed-420a-b9bb-d949829c13fd)

### Steps to reproduce

`nvim -u init.lua`:

```lua
vim.opt.cursorcolumn = true

local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({
    "git",
    "clone",
    "--filter=blob:none",
    "https://github.com/folke/lazy.nvim.git",
    "--branch=stable",
    lazypath,
  })
end
vim.opt.rtp:prepend(lazypath)


local plugins = {
	{
		"nvim-treesitter/nvim-treesitter",
		build = ":TSUpdate",
		event = { "BufReadPost", "BufNewFile" },
		opts = {
			highlight = {
				enable = true,
			},
			indent = { enable = true },
			ensure_installed = "typescript",
			sync_install = false,
			auto_install = true,
		},
		config = function(_, opts)
			require("nvim-treesitter.configs").setup(opts)
		end,
	},
}

require("lazy").setup(plugins)
```

`repro.ts` (sorry, I couldn't get a super clear effect with <100 LOC - just paste it in, press O and start typing):
```typescript
export function findClosestFoundryToml(
    directory: string | undefined,
): string | null {
    let currentDir = directory;
    const root = path.parse(directory).root;

    while (currentDir !== root) {
        const tomlPath = path.join(currentDir, "foundry.toml");
        if (fs.existsSync(tomlPath)) {
            return currentDir;
        }
        const nextDir = path.dirname(currentDir);
        if (nextDir.length >= currentDir.length) break;
        currentDir = nextDir;
    }
    return null;
}

async function runFoundry(
    testData: FoundryTestData,
    item: vscode.TestItem,
    run: vscode.TestRun,
): Promise<string> {
    const { filename, contractName, testName } = testData;
    const verbosity =
        vscode.workspace.getConfiguration("foundryTestRunner").verbosity;
    const cwd = filename && findClosestFoundryToml(item.uri?.path);

    let command = `forge test ${verbosity}`;
    if (cwd) command = `cd ${cwd} && ${command}`;
    if (filename)
        command += ` --match-path "**/${filename}${
            filename.endsWith(".t.sol") ? "" : "/**"
        }"`;
    if (contractName) command += ` --match-contract ${contractName}`;
    if (testName) command += ` --match-test ${testName}`;

    let ret: { stdout: string; stderr: string };
    let code = 0;
    try {
        ret = await promisify(cp.exec)(command);
    } catch (e) {
        ret = e as { stdout: string; stderr: string };
        code = (e as { code: number }).code;
    }
    const { stdout, stderr } = ret;

    run.appendOutput(`Foundry command: ${command}\r\n`);
    run.appendOutput(`Working directory: ${process.cwd()}\r\n`);
    run.appendOutput(unixToDos(stdout));
    run.appendOutput(unixToDos(stderr));
    if (code !== 0) {
        run.appendOutput(`Foundry exited with code: ${code}\r\n`);
    }

    return stdout;
}
export class TestFile extends FoundryTestData {
    /**
     * Parses the tests from the input text, and updates the tests contained
     * by this file to be those from the text,
     */
    public updateFromContents(
        controller: vscode.TestController,
        content: string,
        item: vscode.TestItem,
    ) {
        parseSolidity(content, {
            onTest: (range, testName) => {
                const data = new TestCase(
                    testName,
                    currentContractName,
                    filename,
                    thisGeneration,
                );
                const id = `${item.uri}/${testName}`;
                const tcase = controller.createTestItem(id, testName, item.uri);
                testData.set(tcase, data);
                tcase.range = range;
                parent?.children.add(tcase);
            },

            onContract: (range, contractName) => {
                const id = `${item.uri}/${contractName}`;

                const tcontract = controller.createTestItem(
                    id,
                    contractName,
                    item.uri,
                );
                parent = tcontract;
                currentContractName = contractName;
                tcontract.range = range;
                testData.set(
                    tcontract,
                    new TestContract(contractName, filename, thisGeneration),
                );
                item.children.add(tcontract);
            },
        });
    }
}
```

### Expected behavior

no performance issues as in `2ca076e45^`

### Neovim version (nvim -v)

NVIM v0.10.0-dev-814+g2ca076e45

### Vim (not Nvim) behaves the same?

no

### Operating system/version

macOS 13.4.1

### Terminal name/version

kitty 0.28.1

### $TERM environment variable

xterm-kitty

### Installation

build from repo

**Comment by lucario387** [comment]
nvim-treesitter's indent reparses basically all regions, highlighting I don't think is the issue here.

Related: nvim-treesitter/nvim-treesitter#5347

**Comment by lewis6991** [comment]
The problem is that indent and highlighting are fighting with each other.

When highlight does an update, it does a partial parse of the tree but also invalidates all injections. And then when indent runs it does a complete parse of all the trees.

This is a problem I'm already aware about, and is something I plan to address at some point.

**Comment by llllvvuu** [comment]
> When highlight does an update, it does a partial parse of the tree but also invalidates all injections. And then when indent runs it does a complete parse of all the trees.

I think it's more than just full-reparsing the tree. There is an actual infinite recursion that I just found. I've updated the issue description.

**Comment by lewis6991** [comment]
Isn't that the injection query that is running on every parse? This is why `parse(true)` is so inefficient and why nvim-treesitter-indent should avoid using it.

**Comment by llllvvuu** [comment]
> Isn't that the injection query that is running on every parse? This is why `parse(true)` is so inefficient and why nvim-treesitter-indent should avoid using it.

Having that query run more often isn't enough to explain it, since `parse(true)` actually fixes things. There seems to be an additional issue where the injected query is running on the parent.

i.e. the sexpr that I just put in the issue description should only run inside a comment, but it's running on the whole source file and matching every token. `parse(true)` by itself isn't enough to explain 2600 matches, it would still get <5 matches, just it would run more often.

**Comment by lewis6991** [comment]
Interesting, sounds like this needs to be reduced down much further then.

**Comment by llllvvuu** [comment]
> Interesting, sounds like this needs to be reduced down much further then.

Yes, you're right, I had a lot of wrong leads in this issue, I'll archive this investigation and open a cleaner issue.

**Comment by towry** [comment]
![截屏2023-09-11 11 04 11](https://github.com/neovim/neovim/assets/8279858/7b4e891e-7617-4caa-bd49-58903a0711a7)


I am using `nvim-treesitter` main branch and nightly build neovim, I have to disable highlight to avoid this slowness.

**Comment by llllvvuu** [comment]
@towry it's ok, @lucario387 created fix yesterday! now that i look back it is just breaking standard API breakage; it just felt wonky because the effects manifested in a strange way and i confused myself with false leads.

fix can be backported to master since both the old and new API behave correctly when range is passed in (Lua ignores extra arguments):
https://github.com/llllvvuu/nvim-treesitter/commit/4e2e56dee9b50d9c45d6712565367d79832e2a43

**Comment by towry** [comment]
https://github.com/neovim/neovim/assets/8279858/4363910e-acbb-4a3c-ad0c-3144b8ba4c63

nvim-treestiter#main latest not working for me. I will try to create a minimal config to repro it.

**Comment by lucario387** [comment]
The range for re-parsed of indents is not the best range that we can get, but it's the quickest we can get without needing costly calculation (only visible lines).

**Comment by llllvvuu** [comment]
@towry I actually did manage to reproduce this with no plugins and 4 lines in init.lua. It's *really* elusive though, for example the repro works in my shell, but not when I launch tmux from the same shell. I'll create a new issue though.

**Comment by llllvvuu** [comment]
@towry IDK if it's related to this issue at all, but it may be related to what you're experiencing: #25104

**Comment by towry** [comment]
> 2023-09-11.17.18.49.mov 
> nvim-treestiter#main latest not working for me. I will try to create a minimal config to repro it.

@llllvvuu this issue is not related to current issue or #25104. 

I have created a minimal config to repro it https://gist.github.com/towry/74d6ff3312e18c6fd39367d1b4e477cf

Should I create issue on neovim or nvim-treesitter, what do you suggest, thanks.

**Comment by towry** [comment]
> Have you tested with the latest Neovim `master`?

yes, and tested on kitty and macos terminal.

**Comment by llllvvuu** [comment]
@towry I can reproduce on master, thanks. re:

> Should I create issue on neovim or nvim-treesitter, what do you suggest, thanks.

I think if

```lua
vim.opt.indentexpr = [[v:lua.require('nvim-treesitter').indentexpr()]]
```

is critical to the repro, then you should report in nvim-treesitter. Otherwise everything else seems like it could be done in stock Neovim.

**Comment by towry** [comment]
removed `vim.opt.indentexpr` from the config, remove `vim.treesitter.start` can avoid this slowness.
